openmls/src/group/mls_group/validation.rs

```diff
@@ -0,0 +1,244 @@
+//! This module contains validation functions for incoming messages
+//! as defined in https://github.com/openmls/openmls/wiki/Message-validation
+
+use std::collections::HashSet;
+
+use super::{proposals::StagedProposalQueue, *};
+
+impl MlsGroup {
+    // === Messages ===
+
+    /// Checks the following semantic validation:
+    ///  - ValSem2
+    ///  - ValSem3
+    pub fn validate_framing(&self, message: &MlsMessageIn) -> Result<(), MlsGroupError> {
+        // ValSem2
+        if message.group_id() != self.group_id() {
+            return Err(FramingValidationError::WrongGroupId.into());
+        }
+
+        // ValSem3
+        if message.epoch() != self.context().epoch() {
+            return Err(FramingValidationError::WrongEpoch.into());
+        }
+
+        Ok(())
+    }
+
+    /// Checks the following semantic validation:
+    ///  - ValSem4
+    ///  - ValSem5
+    ///  - ValSem7
+    ///  - ValSem9
+    pub fn validate_plaintext(
+        &self,
+        plaintext: &VerifiableMlsPlaintext,
+    ) -> Result<(), MlsGroupError> {
+        // ValSem4
+        let sender = plaintext.sender();
+        if sender.is_member() {
+            let sender_index = sender.to_leaf_index();
+            if sender_index > self.tree().leaf_count() || self.tree().nodes[sender_index].is_blank()
```

Comment:
In the PR description, ValSem4 is described with "checking that the member exists". This brings us back to the discussion of what a member is. Here, we're checking if the sender's claimed leaf index is within the bounds of the tree. Does that mean that "the member exists"? This is a somewhat philosophical question, but I think what I'm trying to say is that there could be a better label for this check 😄

Reply:
I changed the label in the wiki.